source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -218,12 +219,239 @@ export class ValidationResults {
     }
 }
 
+export abstract class PropertyDefinition {
+    abstract parse(value: any): any;
+    abstract toJSON(target: object, value: any);
+
+    constructor(
+        readonly targetVersion: Shared.TargetVersion,
+        readonly name: string,
+        readonly defaultValue?: any) { }
+}
+
+export abstract class TypedPropertyDefinition<T> extends PropertyDefinition {
+    constructor(
+        readonly targetVersion: Shared.TargetVersion,
+        readonly name: string,
+        readonly defaultValue?: T) {
+        super(targetVersion, name, defaultValue);
+    }
+}
+
+export class StringPropertyDefinition extends TypedPropertyDefinition<string> {
+    parse(value: any): any {
+        let parsedValue = Utils.getStringValue(value, this.defaultValue);
+        let isUndefined = parsedValue === undefined || (parsedValue === "" && this.treatEmptyAsUndefined);
+
+        if (!isUndefined && this.regEx !== undefined) {
+            let matches = this.regEx.exec(value);
+
+            if (!matches) {
+                // TODO: Log parse error
+                return undefined;
+            }
+        }
+
+        return parsedValue;
+    }
+
+    toJSON(target: object, value: any) {
+        Utils.setProperty(
+            target,
+            this.name,
+            value === "" && this.treatEmptyAsUndefined ? undefined : value,
+            this.defaultValue);
+    }
+
+    constructor(
+        readonly targetVersion: Shared.TargetVersion,
+        readonly name: string,
+        readonly treatEmptyAsUndefined: boolean = true,
+        readonly regEx?: RegExp,
+        readonly defaultValue?: string) {
+        super(targetVersion, name, defaultValue);
+    }
+}
+
+export class BooleanPropertyDefinition extends TypedPropertyDefinition<boolean> {
+    parse(value: any): any {
+        return Utils.getBoolValue(value, this.defaultValue);;
+    }
+
+    toJSON(target: object, value: any) {
+        Utils.setProperty(
+            target,
+            this.name,
+            value,
+            this.defaultValue);
+    }
+}
+
+export interface IVersionedValue {
+    targetVersion: Shared.TargetVersion;
+    value: string;
+}
+
+export class ValueSetPropertyDefinition extends TypedPropertyDefinition<string> {
+    parse(value: any): any {
+        let parsedValue = Utils.getStringValue(value, this.defaultValue);
+
+        for (let value of this.values) {
+            if (parsedValue.toLowerCase() === value.value) {
```

Comment:
>parsedValue.toLowerCase() [](start = 16, length = 25)

might be a little more efficient to calculate the lowercase version once. perhaps 

```
let parsedValue = Utils.getStringValue(value, this.defaultValue).toLowerCase();
```

Reply:
This is outdated. BTW I don't know if your using CodeFlow or GitHub for this review, but GitHub has an "outdated" indicator that makes skipping over these easier:
![image](https://user-images.githubusercontent.com/1334689/66151819-edf73100-e5cc-11e9-9294-90107b2958f4.png)

Reply:
unfortunately, that's not what I see:

![image](https://user-images.githubusercontent.com/16614499/66418621-54a69100-e9b7-11e9-9d41-b3a3b3308bfd.png)

Reply:
Weird...